torch/_higher_order_ops/triton_kernel_wrap.py

```diff
@@ -931,3 +931,156 @@ def call_triton_kernel(self, variable, args, kwargs, tx):
+    The ``capture_triton`` API returns a new callable that can actually
+    be traced into a graph.
+
+    Examples:
+
+        >>> import torch
+        >>> import triton
+        >>> from triton import language as tl
+        >>> from torch.fx.experimental.proxy_tensor import make_fx
+        >>> from torch._higher_order_ops.triton_kernel_wrap import capture_triton
+        >>>
+        >>> @triton.jit
+        >>> def add_kernel(
+        >>>     in_ptr0,
+        >>>     in_ptr1,
+        >>>     out_ptr,
+        >>>     n_elements,
+        >>>     BLOCK_SIZE: "tl.constexpr",
+        >>> ):
+        >>>     pid = tl.program_id(axis=0)
+        >>>     block_start = pid * BLOCK_SIZE
+        >>>     offsets = block_start + tl.arange(0, BLOCK_SIZE)
+        >>>     mask = offsets < n_elements
+        >>>     x = tl.load(in_ptr0 + offsets, mask=mask)
+        >>>     y = tl.load(in_ptr1 + offsets, mask=mask)
+        >>>     output = x + y
+        >>>     tl.store(out_ptr + offsets, output, mask=mask)
+        >>>
+        >>> def add(x, y):
+        >>>     output = torch.empty_like(x)
+        >>>     n_elements = output.numel()
+        >>>
+        >>>     def grid_fn(meta):
+        >>>         return (triton.cdiv(n_elements, meta["BLOCK_SIZE"]),)
+        >>>
+        >>>     capture_triton(add_kernel)[grid_fn](x, y, output, n_elements, 16)
+        >>>     return output
+        >>>
+        >>> x = torch.randn(3, device="cuda")
+        >>> y = torch.randn(3, device="cuda")
+        >>> gm = make_fx(add)(x, y)
+        >>> print(gm.code)
+        >>> # def forward(self, x_1, y_1):
+        >>> #     empty_like = torch.ops.aten.empty_like.default(x_1, pin_memory = False)
+        >>> #     triton_kernel_wrapper_mutation_proxy = triton_kernel_wrapper_mutation(
+        >>> #         kernel_idx = 0, constant_args_idx = 0,
+        >>> #         grid = [(1, 1, 1)], kwargs = {
+        >>> #             'in_ptr0': x_1, 'in_ptr1': y_1, 'out_ptr': empty_like,
+        >>> #             'n_elements': 3, 'BLOCK_SIZE': 16
+        >>> #         })
+        >>> #     return empty_like
+
+    """
+    from triton.runtime.autotuner import Autotuner
+    from triton.runtime.jit import JITFunction
+
+    if not isinstance(triton_kernel, (JITFunction, Autotuner)):
+        raise RuntimeError(
+            "capture_triton only works on functions annotated with triton.jit or triton.autotune"
+        )
+    return TraceableTritonKernelWrapper(triton_kernel, None, None)
+
+
+from ..fx._symbolic_trace import is_fx_tracing
+
+
+class TracingTritonHOPifier(TritonHOPifier):
+    def raise_unsupported(self, msg):
+        raise RuntimeError(msg)
+
+    def is_callable(self, maybe_callable):
+        return callable(maybe_callable)
+
+    def call_grid(self, grid, meta, tx):
+        assert tx is None
+        return grid(meta)
+
+    def check_grid(self, grid):
+        if not isinstance(grid, tuple):
```

Comment:
1) I haven't actually verified this but is it really only ever tuple? would say a list fail here? IIRC triton only reads it as decomposition, but good to verify

2) not sure how the orchestration here happens but if grid is a function, we do already have it fully resolved before here?

Reply:
> 1. I haven't actually verified this but is it really only ever tuple? would say a list fail here? IIRC triton only reads it as decomposition, but good to verify

A list works here, so I'll update the code to convert to tuple (and add a test for it)

> 2. not sure how the orchestration here happens but if grid is a function, we do already have it fully resolved before here? can you point to who calls these?

If grid is a function, it's fully resolved before here. The orchestrator resolves the function before calling check_grid; the code for this is over at https://github.com/pytorch/pytorch/pull/130178/files#diff-f5fa7d0e418e91c63fa56d577a92a294c87e19318a3c8b3736ac4254eaa51db9R907-R915